# Task
Find and display the common links between the 'restaurant_yelp_links.csv' and 'pending_ipo_restaurants_dec2.csv' datasets, then summarize the number of matching links found.

## Load Data

### Subtask:
Load 'restaurant_yelp_links.csv' and 'pending_ipo_restaurants_dec2.csv' into separate pandas DataFrames.


**Reasoning**:
Load the two specified CSV files into pandas DataFrames and then display their first 5 rows and column names to inspect their structure and content, as per the instructions.



In [ ]:
import pandas as pd

df_yelp = pd.read_csv('/content/restaurant_yelp_links.csv')
df_ipo = pd.read_csv('/content/pending_ipo_resaurants_dec2.csv')

print("df_yelp Head:\n")
print(df_yelp.head())
print("\n")
print("df_yelp Columns:\n")
print(df_yelp.columns)
print("\n")

print("df_ipo Head:\n")
print(df_ipo.head())
print("\n")
print("df_ipo Columns:\n")
print(df_ipo.columns)
print("\n")

df_yelp Head:

                                          Restaurant  \
0          CAVA – Webster, TX (1501 C Bay Area Blvd)   
1  First Watch – League City, TX (1720 W FM 646 S...   
2  GEN Korean BBQ House – Houston, TX (18299 Egre...   
3   Portillo’s – Richmond, TX (11005 W Grand Pkwy S)   
4      CAVA – Huntington Beach, CA (10054 Adams Ave)   

                     Location  \
0           Webster, TX 77598   
1       League City, TX 77573   
2           Houston, TX 77058   
3          Richmond, TX 77407   
4  Huntington Beach, CA 92646   

                                           Yelp Link  
0              https://www.yelp.com/biz/cava-webster  
1  https://www.yelp.com/biz/first-watch-league-ci...  
2  https://www.yelp.com/biz/gen-korean-bbq-house-...  
3  https://www.yelp.com/biz/portillos-richmond-ri...  
4     https://www.yelp.com/biz/cava-huntington-beach  


df_yelp Columns:

Index(['Restaurant', 'Location', 'Yelp Link'], dtype='object')


df_ipo Head:

  Restaurant Chain  

## Identify Common Links

### Subtask:
Find the links that are present in both DataFrames. This will involve inspecting the columns of both datasets to determine the appropriate column(s) for comparison.


**Reasoning**:
To compare the links effectively, I need to standardize them by removing any query parameters. I'll define a function to extract the base URL and then apply it to the 'Yelp Link' column in `df_yelp` and the 'Sample Review' column in `df_ipo` to create new 'base_link' columns.



In [ ]:
from urllib.parse import urlparse, urlunparse

def get_base_url(url):
    """Extracts the base URL without query parameters."""
    if pd.isna(url): # Handle potential NaN values
        return None
    parsed_url = urlparse(url)
    return urlunparse(parsed_url._replace(query='', fragment=''))

# Apply the function to df_yelp['Yelp Link']
df_yelp['yelp_base_link'] = df_yelp['Yelp Link'].apply(get_base_url)

# Apply the function to df_ipo['Sample Review']
df_ipo['ipo_base_link'] = df_ipo['Sample Review'].apply(get_base_url)

print("df_yelp with new 'yelp_base_link' column:")
print(df_yelp[['Yelp Link', 'yelp_base_link']].head())
print("\ndf_ipo with new 'ipo_base_link' column:")
print(df_ipo[['Sample Review', 'ipo_base_link']].head())

df_yelp with new 'yelp_base_link' column:
                                           Yelp Link  \
0              https://www.yelp.com/biz/cava-webster   
1  https://www.yelp.com/biz/first-watch-league-ci...   
2  https://www.yelp.com/biz/gen-korean-bbq-house-...   
3  https://www.yelp.com/biz/portillos-richmond-ri...   
4     https://www.yelp.com/biz/cava-huntington-beach   

                                      yelp_base_link  
0              https://www.yelp.com/biz/cava-webster  
1  https://www.yelp.com/biz/first-watch-league-ci...  
2  https://www.yelp.com/biz/gen-korean-bbq-house-...  
3  https://www.yelp.com/biz/portillos-richmond-ri...  
4     https://www.yelp.com/biz/cava-huntington-beach  

df_ipo with new 'ipo_base_link' column:
                                       Sample Review  \
0   https://www.yelp.com/biz/cava-anaheim-3?start=60   
1    https://www.yelp.com/biz/cava-dallas-3?start=40   
2  https://www.yelp.com/biz/cava-flower-mound?sta...   
3        https://www.yelp.

**Reasoning**:
Now that the base URLs have been extracted and standardized, I will find the common links by taking the intersection of the unique 'yelp_base_link' values from `df_yelp` and the unique 'ipo_base_link' values from `df_ipo`. Finally, I will print the count of common links found.



In [ ]:
common_links = pd.Series(list(set(df_yelp['yelp_base_link'].dropna()) & set(df_ipo['ipo_base_link'].dropna())))

print(f"Number of common links found: {len(common_links)}")

if not common_links.empty:
    print("\nFirst 5 common links:")
    print(common_links.head())
else:
    print("\nNo common links found.")

Number of common links found: 4

First 5 common links:
0    https://www.yelp.com/biz/first-watch-league-ci...
1    https://www.yelp.com/biz/gen-korean-bbq-house-...
2              https://www.yelp.com/biz/cava-anaheim-3
3    https://www.yelp.com/biz/portillos-richmond-ri...
dtype: object


## Display and Summarize Common Links

### Subtask:
Display all common links and summarize the findings.


**Reasoning**:
To fulfill the subtask, I will display all the common links and then summarize the findings, including the total count and additional insights, by first printing all common links using a code block and then providing a detailed summary in a markdown block.



In [ ]:
print("\nAll common links found:")
for link in common_links:
    print(link)


All common links found:
https://www.yelp.com/biz/first-watch-league-city-6
https://www.yelp.com/biz/gen-korean-bbq-house-houston-10
https://www.yelp.com/biz/cava-anaheim-3
https://www.yelp.com/biz/portillos-richmond-richmond-2


A total of **4** common Yelp links were identified between the 'restaurant_yelp_links.csv' and 'pending_ipo_restaurants_dec2.csv' datasets. This indicates that these 4 restaurant locations are present in both datasets, signifying a direct match in their Yelp presence. These links represent the intersection of businesses for which Yelp information is available in both the general Yelp links dataset and the pending IPO restaurants dataset.

## Summary:

### Q&A
The number of matching common Yelp links found between the 'restaurant\_yelp\_links.csv' and 'pending\_ipo\_restaurants\_dec2.csv' datasets is 4.

### Data Analysis Key Findings
*   Two datasets, 'restaurant\_yelp\_links.csv' (df\_yelp) and 'pending\_ipo\_restaurants\_dec2.csv' (df\_ipo), were successfully loaded.
    *   `df_yelp` contained `Restaurant`, `Location`, and `Yelp Link` columns.
    *   `df_ipo` contained `Restaurant Chain`, `Sample Review` (identified as Yelp links), and other related information.
*   Yelp links from both datasets were standardized by extracting their base URL, removing query parameters (e.g., `?start=XX`). This involved creating new columns: `yelp_base_link` for `df_yelp` and `ipo_base_link` for `df_ipo`.
*   A total of **4 common Yelp links** were identified by finding the intersection of these standardized base URLs between the two datasets.
*   The specific common links found were:
    *   `https://www.yelp.com/biz/first-watch-league-city-6`
    *   `https://www.yelp.com/biz/gen-korean-bbq-house-houston-10`
    *   `https://www.yelp.com/biz/cava-anaheim-3`
    *   `https://www.yelp.com/biz/portillos-richmond-richmond-2`

### Insights or Next Steps
*   These 4 common links represent specific restaurant locations present in both the general Yelp links dataset and the pending IPO restaurants dataset, indicating a direct overlap in their Yelp presence. This shared data could be used for cross-referencing information or enriching one dataset with details from the other.
*   Further analysis could involve investigating the restaurant names associated with these common links to understand which specific chains or locations are consistently tracked across both data sources, and exploring why only a small number of overlaps exist.


**Reasoning**:
To extract the links for the specified restaurant chains, I will filter `df_ipo` for each chain using the 'Restaurant Chain' column and then display the unique 'ipo_base_link' for the matching entries.

In [ ]:
import pandas as pd
import os
from urllib.parse import urlparse, urlunparse

def get_base_url(url):
    """Extracts the base URL without query parameters."""
    if pd.isna(url):
        return None
    parsed_url = urlparse(url)
    return urlunparse(parsed_url._replace(query='', fragment=''))

# Check if dataframes are already loaded to avoid unnecessary re-loading
if 'df_yelp' not in locals() or 'df_ipo' not in locals() or df_yelp.empty or df_ipo.empty:
    print("DataFrames not found or empty. Attempting to reload...")

    # Attempt to load df_yelp
    try:
        df_yelp = pd.read_csv('/content/restaurant_yelp_links.csv')
        df_yelp['yelp_base_link'] = df_yelp['Yelp Link'].apply(get_base_url)
        print("df_yelp loaded successfully.")
    except FileNotFoundError:
        print("Warning: '/content/restaurant_yelp_links.csv' not found. Some previous steps might be affected.")
        df_yelp = pd.DataFrame() # Create an empty DataFrame to avoid errors if not critical for current task

    # Attempt to load df_ipo (try CSV first, then XLSX)
    try:
        df_ipo = pd.read_csv('/content/pending_ipo_resaurants_dec2.csv')
        print("df_ipo (CSV) loaded successfully.")
    except FileNotFoundError:
        print("'/content/pending_ipo_resaurants_dec2.csv' not found. Trying '/content/pending_ipo_resaurants_dec2.xlsx'...")
        try:
            df_ipo = pd.read_excel('/content/pending_ipo_resaurants_dec2.xlsx')
            print("df_ipo (Excel) loaded successfully.")
        except FileNotFoundError:
            print("Error: Neither '/content/pending_ipo_resaurants_dec2.csv' nor '.xlsx' found.")
            df_ipo = pd.DataFrame() # Create an empty DataFrame to avoid errors

    if not df_ipo.empty:
        df_ipo['ipo_base_link'] = df_ipo['Sample Review'].apply(get_base_url)
    else:
        print("Cannot proceed with df_ipo related tasks as the data is missing.")

else:
    print("DataFrames already loaded, proceeding with existing data.")

# List of restaurant chains to extract links for
restaurant_chains_to_extract = [
    'CAVA', 'Krispy Kreme', 'Sweetgreen', 'GEN Korean BBQ', 'First Watch', 'Kura Sushi',
    'Reborn Coffee', 'Yoshiharu Ramen', 'Fatburger', 'Portillo\'s'
]

if not df_ipo.empty:
    print("\nExtracting ipo_base_links for specified restaurant chains from df_ipo:\n")

    for chain in restaurant_chains_to_extract:
        # Filter df_ipo for the current restaurant chain
        filtered_df = df_ipo[df_ipo['Restaurant Chain'] == chain]

        # Get unique ipo_base_links for the filtered chain
        links = filtered_df['ipo_base_link'].dropna().unique()

        print(f"--- {chain} ({len(links)} links) ---")
        if len(links) > 0:
            for link in links:
                print(link)
        else:
            print("No links found for this chain.")
        print("\n")
else:
    print("Cannot extract links as df_ipo could not be loaded.")

DataFrames not found or empty. Attempting to reload...
'/content/pending_ipo_resaurants_dec2.csv' not found. Trying '/content/pending_ipo_resaurants_dec2.xlsx'...
df_ipo (Excel) loaded successfully.

Extracting ipo_base_links for specified restaurant chains from df_ipo:

--- CAVA (5 links) ---
https://www.yelp.com/biz/cava-anaheim-3
https://www.yelp.com/biz/cava-dallas-3
https://www.yelp.com/biz/cava-flower-mound
https://www.yelp.com/biz/cava-san-antonio-4
https://www.yelp.com/biz/zoes-kitchen-houston-39


--- Krispy Kreme (5 links) ---
https://www.yelp.com/biz/krispy-kreme-allen
https://www.yelp.com/biz/krispy-kreme-el-paso-4
https://www.yelp.com/biz/krispy-kreme-la-habra-2
https://www.yelp.com/biz/krispy-kreme-mission-viejo
https://www.yelp.com/biz/krispy-kreme-redlands


--- Sweetgreen (4 links) ---
https://www.yelp.com/biz/sweetgreen-san-diego-6
https://www.yelp.com/biz/sweetgreen-santa-rosa
https://www.yelp.com/biz/sweetgreen-west-hollywood-2
https://www.yelp.com/biz/sweetgreen-lo

**Reasoning**:
To extract the links for 'Dutch Bros' locations in Texas, I will filter `df_ipo` where 'Restaurant Chain' is 'Dutch Bros' and 'State' is 'Texas', then display the unique 'ipo_base_link' for these filtered entries.

In [ ]:
import pandas as pd
from urllib.parse import urlparse, urlunparse

def get_base_url(url):
    """Extracts the base URL without query parameters."""
    if pd.isna(url):
        return None
    parsed_url = urlparse(url)
    return urlunparse(parsed_url._replace(query='', fragment=''))

# Ensure df_ipo is loaded and ipo_base_link is created
# This block is included for robustness in case the kernel state is lost, similar to previous cells
if 'df_ipo' not in locals() or df_ipo.empty:
    try:
        df_ipo = pd.read_csv('/content/pending_ipo_resaurants_dec2.csv')
        print("df_ipo (CSV) loaded successfully.")
    except FileNotFoundError:
        try:
            df_ipo = pd.read_excel('/content/pending_ipo_resaurants_dec2.xlsx')
            print("df_ipo (Excel) loaded successfully.")
        except FileNotFoundError:
            print("Error: Neither '/content/pending_ipo_resaurants_dec2.csv' nor '.xlsx' found.")
            df_ipo = pd.DataFrame() # Create an empty DataFrame to avoid errors

    if not df_ipo.empty:
        df_ipo['ipo_base_link'] = df_ipo['Sample Review'].apply(get_base_url)
    else:
        print("Cannot proceed as df_ipo could not be loaded.")


if not df_ipo.empty:
    print("\nExtracting ipo_base_links for Dutch Bros in Texas from df_ipo:\n")

    # Filter df_ipo for 'Dutch Bros' and 'Texas'
    filtered_dutchbros_texas = df_ipo[
        (df_ipo['Restaurant Chain'] == 'Dutch Bros') &
        (df_ipo['State'] == 'Texas')
    ]

    # Get unique ipo_base_links for the filtered chain
    links = filtered_dutchbros_texas['ipo_base_link'].dropna().unique()

    print(f"--- Dutch Bros in Texas ({len(links)} links) ---")
    if len(links) > 0:
        for link in links:
            print(link)
    else:
        print("No links found for Dutch Bros in Texas.")
    print("\n")
else:
    print("Cannot extract links as df_ipo could not be loaded.")

df_ipo (Excel) loaded successfully.

Extracting ipo_base_links for Dutch Bros in Texas from df_ipo:

--- Dutch Bros in Texas (70 links) ---
https://www.yelp.com/biz/dutch-bros-coffee-coming-soon-el-paso
https://www.yelp.com/biz/dutch-bros-coffee-opening-soon-wylie
https://www.yelp.com/biz/dutch-bros-coffee-corsicana-2
https://www.yelp.com/biz/dutch-bros-coffee-amarillo-4
https://www.yelp.com/biz/dutch-bros-coffee-missouri-city
https://www.yelp.com/biz/dutch-bros-coffee-houston-3
https://www.yelp.com/biz/dutch-bros-coffee-san-marcos-2
https://www.yelp.com/biz/dutch-bros-coffee-buda-2
https://www.yelp.com/biz/dutch-bros-coffee-dripping-springs-2
https://www.yelp.com/biz/dutch-bros-coffee-deer-park
https://www.yelp.com/biz/dutch-bros-coffee-la-porte
https://www.yelp.com/biz/dutch-bros-coffee-katy-2
https://www.yelp.com/biz/dutch-bros-coffee-mont-belvieu
https://www.yelp.com/biz/dutch-bros-coffee-baytown-2
https://www.yelp.com/biz/dutch-bros-coffee-baytown-3
https://www.yelp.com/biz/dutch-

**Reasoning**:
To extract the links for 'Sweet Green' locations in Texas, I will filter `df_ipo` where 'Restaurant Chain' is 'Sweet Green' and 'State' is 'Texas', then display the unique 'ipo_base_link' for these filtered entries.

In [ ]:
import pandas as pd
from urllib.parse import urlparse, urlunparse

def get_base_url(url):
    """Extracts the base URL without query parameters."""
    if pd.isna(url):
        return None
    parsed_url = urlparse(url)
    return urlunparse(parsed_url._replace(query='', fragment=''))

# Ensure df_ipo is loaded and ipo_base_link is created
if 'df_ipo' not in locals() or df_ipo.empty:
    try:
        df_ipo = pd.read_csv('/content/pending_ipo_resaurants_dec2.csv')
        print("df_ipo (CSV) loaded successfully.")
    except FileNotFoundError:
        try:
            df_ipo = pd.read_excel('/content/pending_ipo_resaurants_dec2.xlsx')
            print("df_ipo (Excel) loaded successfully.")
        except FileNotFoundError:
            print("Error: Neither '/content/pending_ipo_resaurants_dec2.csv' nor '.xlsx' found.")
            df_ipo = pd.DataFrame() # Create an empty DataFrame to avoid errors

    if not df_ipo.empty:
        df_ipo['ipo_base_link'] = df_ipo['Sample Review'].apply(get_base_url)
    else:
        print("Cannot proceed as df_ipo could not be loaded.")


if not df_ipo.empty:
    print("\nExtracting ipo_base_links for Sweet Green in Texas from df_ipo:\n")

    # Filter df_ipo for 'Sweet Green' and 'Texas'
    filtered_sweetgreen_texas = df_ipo[
        (df_ipo['Restaurant Chain'] == 'Sweet Green') &
        (df_ipo['State'] == 'Texas')
    ]

    # Get unique ipo_base_links for the filtered chain
    links = filtered_sweetgreen_texas['ipo_base_link'].dropna().unique()

    print(f"--- Sweet Green in Texas ({len(links)} links) ---")
    if len(links) > 0:
        for link in links:
            print(link)
    else:
        print("No links found for Sweet Green in Texas.")
    print("\n")
else:
    print("Cannot extract links as df_ipo could not be loaded.")

df_ipo (Excel) loaded successfully.

Extracting ipo_base_links for Sweet Green in Texas from df_ipo:

--- Sweet Green in Texas (19 links) ---
https://www.yelp.com/biz/sweetgreen-hillcrest-ave-dallas
https://www.yelp.com/biz/sweetgreen-houston
https://www.yelp.com/biz/sweetgreen-austin-4
https://www.yelp.com/biz/sweetgreen-houston-14
https://www.yelp.com/biz/sweetgreen-austin-6
https://www.yelp.com/biz/sweetgreen-austin-10
https://www.yelp.com/biz/sweetgreen-houston-5
https://www.yelp.com/biz/sweetgreen-forth-worth
https://www.yelp.com/biz/sweetgreen-houston-3
https://www.yelp.com/biz/sweetgreen-dallas
https://www.yelp.com/biz/sweetgreen-austin-3
https://www.yelp.com/biz/sweetgreen-san-antonio-2
https://www.yelp.com/biz/sweetgreen-austin-2
https://www.yelp.com/biz/sweetgreen-southlake
https://www.yelp.com/biz/sweetgreen-houston-11
https://www.yelp.com/biz/sweetgreen-dallas-7
https://www.yelp.com/biz/sweetgreen-the-woodlands
https://www.yelp.com/biz/sweetgreen-cypress-3
https://www.yelp.

**Reasoning**:
To extract the links for 'Sweet Green' locations in California, I will filter `df_ipo` where 'Restaurant Chain' is 'Sweet Green' and 'State' is 'California', then display the unique 'ipo_base_link' for these filtered entries.

In [ ]:
import pandas as pd
from urllib.parse import urlparse, urlunparse

def get_base_url(url):
    """Extracts the base URL without query parameters."""
    if pd.isna(url):
        return None
    parsed_url = urlparse(url)
    return urlunparse(parsed_url._replace(query='', fragment=''))

# Ensure df_ipo is loaded and ipo_base_link is created
if 'df_ipo' not in locals() or df_ipo.empty:
    try:
        df_ipo = pd.read_csv('/content/pending_ipo_resaurants_dec2.csv')
        print("df_ipo (CSV) loaded successfully.")
    except FileNotFoundError:
        try:
            df_ipo = pd.read_excel('/content/pending_ipo_resaurants_dec2.xlsx')
            print("df_ipo (Excel) loaded successfully.")
        except FileNotFoundError:
            print("Error: Neither '/content/pending_ipo_resaurants_dec2.csv' nor '.xlsx' found.")
            df_ipo = pd.DataFrame() # Create an empty DataFrame to avoid errors

    if not df_ipo.empty:
        df_ipo['ipo_base_link'] = df_ipo['Sample Review'].apply(get_base_url)
    else:
        print("Cannot proceed as df_ipo could not be loaded.")


if not df_ipo.empty:
    print("\nExtracting ipo_base_links for Sweet Green in California from df_ipo:\n")

    # Filter df_ipo for 'Sweet Green' and 'California'
    filtered_sweetgreen_california = df_ipo[
        (df_ipo['Restaurant Chain'] == 'Sweet Green') &
        (df_ipo['State'] == 'California')
    ]

    # Get unique ipo_base_links for the filtered chain
    links = filtered_sweetgreen_california['ipo_base_link'].dropna().unique()

    print(f"--- Sweet Green in California ({len(links)} links) ---")
    if len(links) > 0:
        for link in links:
            print(link)
    else:
        print("No links found for Sweet Green in California.")
    print("\n")
else:
    print("Cannot extract links as df_ipo could not be loaded.")


Extracting ipo_base_links for Sweet Green in California from df_ipo:

--- Sweet Green in California (47 links) ---
https://www.yelp.com/biz/sweetgreen-newport-beach
https://www.yelp.com/biz/sweetgreen-san-francisco-3
https://www.yelp.com/biz/sweetgreen-studio-city
https://www.yelp.com/biz/sweetgreen-santa-monica-5
https://www.yelp.com/biz/sweetgreen-los-angeles-12
https://www.yelp.com/biz/sweetgreen-west-hollywood-2
https://www.yelp.com/biz/sweetgreen-encino
https://www.yelp.com/biz/sweetgreen-los-angeles
https://www.yelp.com/biz/sweetgreen-palo-alto-2
https://www.yelp.com/biz/sweetgreen-san-diego-3
https://www.yelp.com/biz/sweetgreen-los-angeles-9
https://www.yelp.com/biz/sweetgreen-torrance-3
https://www.yelp.com/biz/sweetgreen-sunnyvale-3
https://www.yelp.com/biz/sweetgreen-berkeley
https://www.yelp.com/biz/sweetgreen-campbell
https://www.yelp.com/biz/sweetgreen-irvine
https://www.yelp.com/biz/sweetgreen-san-diego
https://www.yelp.com/biz/sweetgreen-san-diego-5
https://www.yelp.com

**Reasoning**:
To get all entries with 1 to 10 reviews, I will filter the `sorted_df_control_ca` DataFrame to include only rows where `n_reviews` is greater than or equal to 1 and less than or equal to 10. Then, I will display the 'Outlet', 'url', and 'n_reviews' columns for these filtered entries.

In [ ]:
# Filter sorted_df_control_ca for entries with 1 to 10 reviews
filtered_reviews_1_to_10 = sorted_df_control_ca[
    (sorted_df_control_ca['n_reviews'] >= 1) &
    (sorted_df_control_ca['n_reviews'] <= 10)
]

print("Entries with 1 to 10 reviews:\n")

if not filtered_reviews_1_to_10.empty:
    # Display all entries in this range
    print(filtered_reviews_1_to_10[['Outlet', 'url', 'n_reviews']].to_string(index=False))
else:
    print("No entries found with 1 to 10 reviews.")

Entries with 1 to 10 reviews:

                      Outlet                                                                    url  n_reviews
      Krispy Krunchy Chicken              https://www.yelp.com/biz/krispy-krunchy-chicken-patterson          1
      Krispy Krunchy Chicken                  https://www.yelp.com/biz/krispy-krunchy-chicken-ripon          1
      Krispy Krunchy Chicken                https://www.yelp.com/biz/krispy-krunchy-chicken-turlock          1
      Krispy Krunchy Chicken           https://www.yelp.com/biz/krispy-krunchy-chicken-sacramento-9          1
      Krispy Krunchy Chicken              https://www.yelp.com/biz/krispy-krunchy-chicken-redding-2          1
      Krispy Krunchy Chicken                  https://www.yelp.com/biz/krispy-krunchy-chicken-chico          1
      Krispy Krunchy Chicken                https://www.yelp.com/biz/krispy-krunchy-chicken-yucaipa          1
      Krispy Krunchy Chicken                 https://www.yelp.com/biz/krispy-krun

**Reasoning**:
To provide the list with outlet names kept together, I will re-run the filter on `sorted_df_control_ca` for `n_reviews` between 1 and 10. The `to_string(index=False)` method ensures that multi-word entries in the 'Outlet' column are printed as a single unit, respecting their full name.

In [ ]:
# Filter sorted_df_control_ca for entries with 1 to 10 reviews
filtered_reviews_1_to_10 = sorted_df_control_ca[
    (sorted_df_control_ca['n_reviews'] >= 1) &
    (sorted_df_control_ca['n_reviews'] <= 10)
]

print("Entries with 1 to 10 reviews (Outlet names kept together):\n")

if not filtered_reviews_1_to_10.empty:
    # Display all entries in this range, ensuring Outlet names are kept together
    print(filtered_reviews_1_to_10[['Outlet', 'url', 'n_reviews']].to_string(index=False))
else:
    print("No entries found with 1 to 10 reviews.")

Entries with 1 to 10 reviews (Outlet names kept together):

                      Outlet                                                                    url  n_reviews
      Krispy Krunchy Chicken              https://www.yelp.com/biz/krispy-krunchy-chicken-patterson          1
      Krispy Krunchy Chicken                  https://www.yelp.com/biz/krispy-krunchy-chicken-ripon          1
      Krispy Krunchy Chicken                https://www.yelp.com/biz/krispy-krunchy-chicken-turlock          1
      Krispy Krunchy Chicken           https://www.yelp.com/biz/krispy-krunchy-chicken-sacramento-9          1
      Krispy Krunchy Chicken              https://www.yelp.com/biz/krispy-krunchy-chicken-redding-2          1
      Krispy Krunchy Chicken                  https://www.yelp.com/biz/krispy-krunchy-chicken-chico          1
      Krispy Krunchy Chicken                https://www.yelp.com/biz/krispy-krunchy-chicken-yucaipa          1
      Krispy Krunchy Chicken                 https:/

**Reasoning**:
Now that we have confirmed the presence of the `n_reviews` column in `df_control_ca`, I will sort the DataFrame by this column in ascending order. Then, I will display the 'Outlet', 'url', and 'n_reviews' for the sorted entries to show the list from 0 reviews to the maximum.

In [ ]:
# Sort df_control_ca by 'n_reviews' in ascending order
sorted_df_control_ca = df_control_ca.sort_values(by='n_reviews', ascending=True)

print("Restaurants sorted by number of reviews (0 to max):\n")
# Display the relevant columns for the sorted DataFrame
# Using .head(20) to show a sample, or .to_string() if you want to see all
# For this example, I'll show the top 20 and bottom 5 to illustrate the sorting
print("First 20 entries (lowest review counts):\n")
print(sorted_df_control_ca[['Outlet', 'url', 'n_reviews']].head(20).to_string(index=False))

print("\nLast 5 entries (highest review counts):\n")
print(sorted_df_control_ca[['Outlet', 'url', 'n_reviews']].tail(5).to_string(index=False))

Restaurants sorted by number of reviews (0 to max):

First 20 entries (lowest review counts):

                      Outlet                                                                   url  n_reviews
      Krispy Krunchy Chicken             https://www.yelp.com/biz/krispy-krunchy-chicken-patterson          1
      Krispy Krunchy Chicken                 https://www.yelp.com/biz/krispy-krunchy-chicken-ripon          1
      Krispy Krunchy Chicken               https://www.yelp.com/biz/krispy-krunchy-chicken-turlock          1
      Krispy Krunchy Chicken          https://www.yelp.com/biz/krispy-krunchy-chicken-sacramento-9          1
      Krispy Krunchy Chicken             https://www.yelp.com/biz/krispy-krunchy-chicken-redding-2          1
      Krispy Krunchy Chicken                 https://www.yelp.com/biz/krispy-krunchy-chicken-chico          1
      Krispy Krunchy Chicken               https://www.yelp.com/biz/krispy-krunchy-chicken-yucaipa          1
      Krispy Krunchy Chic

**Reasoning**:
To verify if the provided file `control_restaurants_CA_dec2025.csv` contains the `n_reviews` column, I will load it into a new DataFrame (`df_control_ca`) and display its columns. This will allow us to confirm the column's presence and proceed with the sorting request.

In [ ]:
import pandas as pd

# Load the new CSV file
try:
    df_control_ca = pd.read_csv('/content/control_restaurants_CA_dec2025.csv')
    print("'/content/control_restaurants_CA_dec2025.csv' loaded successfully.")

    print("\nColumns in df_control_ca:")
    print(df_control_ca.columns)

    # Check for 'n_reviews' column
    if 'n_reviews' in df_control_ca.columns:
        print("\n'n_reviews' column found in df_control_ca.")
        # Display first few rows with n_reviews if present (using existing columns)
        if not df_control_ca[['Outlet', 'url', 'n_reviews']].empty:
            print("\nFirst 5 rows of relevant columns from df_control_ca:")
            print(df_control_ca[['Outlet', 'url', 'n_reviews']].head())
        else:
            print("\nNo data to display in relevant columns from df_control_ca.")
    else:
        print("\n'n_reviews' column NOT found in df_control_ca.")
        print("Available columns:", df_control_ca.columns.tolist())

except FileNotFoundError:
    print("Error: '/content/control_restaurants_CA_dec2025.csv' not found.")
    df_control_ca = pd.DataFrame()
except Exception as e:
    print(f"An error occurred while loading or inspecting the file: {e}")
    df_control_ca = pd.DataFrame()

'/content/control_restaurants_CA_dec2025.csv' loaded successfully.

Columns in df_control_ca:
Index(['S.No.', 'Outlet', 'Address', 'ZIP Code', 'url', 'n_reviews',
       'average_rating', 'menu_url', 'categories', 'business_id', 'treated',
       'state'],
      dtype='object')

'n_reviews' column found in df_control_ca.

First 5 rows of relevant columns from df_control_ca:
                     Outlet  \
0             Aromas Coffee   
1             Coffee Cartel   
2            Courtyard Cafe   
3            Courtyard Cafe   
4  Crenshaw Farmerss Market   

                                                 url  n_reviews  
0  https://www.yelp.com/biz/aromas-coffee-los-ang...         24  
1  https://www.yelp.com/biz/coffee-cartel-redondo...        482  
2   https://www.yelp.com/biz/courtyard-cafe-glendale          1  
3  https://www.yelp.com/biz/courtyard-cafe-sherma...          6  
4  https://www.yelp.com/biz/crenshaw-farmers-mark...         27  


**Reasoning**:
To verify the presence of an `n_reviews` column in `df_ipo`, I will display the columns of the DataFrame. This will confirm whether I can use this column for sorting the restaurant links.

In [ ]:
import pandas as pd

# Ensure df_ipo is loaded
if 'df_ipo' not in locals() or df_ipo.empty:
    try:
        df_ipo = pd.read_csv('/content/pending_ipo_resaurants_dec2.csv')
        print("df_ipo (CSV) loaded successfully for column check.")
    except FileNotFoundError:
        try:
            df_ipo = pd.read_excel('/content/pending_ipo_resaurants_dec2.xlsx')
            print("df_ipo (Excel) loaded successfully for column check.")
        except FileNotFoundError:
            print("Error: Neither '/content/pending_ipo_resaurants_dec2.csv' nor '.xlsx' found. Cannot check columns.")
            df_ipo = pd.DataFrame() # Create an empty DataFrame to avoid further errors

if not df_ipo.empty:
    print("Columns in df_ipo:")
    print(df_ipo.columns)

    # Also check if 'n_reviews' is directly in the columns or if there's a similar name
    if 'n_reviews' in df_ipo.columns:
        print("\n'n_reviews' column found in df_ipo.")
    else:
        print("\n'n_reviews' column NOT found in df_ipo. Please check the exact column name if you meant a different column for review counts.")
else:
    print("df_ipo could not be loaded, skipping column check.")

Error: Neither '/content/pending_ipo_resaurants_dec2.csv' nor '.xlsx' found. Cannot check columns.
df_ipo could not be loaded, skipping column check.


**Reasoning**:
To extract the links for 'Dutch Bros' locations in California, I will filter `df_ipo` where 'Restaurant Chain' is 'Dutch Bros' and 'State' is 'California', then display the unique 'ipo_base_link' for these filtered entries.

In [ ]:
import pandas as pd
from urllib.parse import urlparse, urlunparse

def get_base_url(url):
    """Extracts the base URL without query parameters."""
    if pd.isna(url):
        return None
    parsed_url = urlparse(url)
    return urlunparse(parsed_url._replace(query='', fragment=''))

# Ensure df_ipo is loaded and ipo_base_link is created
if 'df_ipo' not in locals() or df_ipo.empty:
    try:
        df_ipo = pd.read_csv('/content/pending_ipo_resaurants_dec2.csv')
        print("df_ipo (CSV) loaded successfully.")
    except FileNotFoundError:
        try:
            df_ipo = pd.read_excel('/content/pending_ipo_resaurants_dec2.xlsx')
            print("df_ipo (Excel) loaded successfully.")
        except FileNotFoundError:
            print("Error: Neither '/content/pending_ipo_resaurants_dec2.csv' nor '.xlsx' found.")
            df_ipo = pd.DataFrame() # Create an empty DataFrame to avoid errors

    if not df_ipo.empty:
        df_ipo['ipo_base_link'] = df_ipo['Sample Review'].apply(get_base_url)
    else:
        print("Cannot proceed as df_ipo could not be loaded.")


if not df_ipo.empty:
    print("\nExtracting ipo_base_links for Dutch Bros in California from df_ipo:\n")

    # Filter df_ipo for 'Dutch Bros' and 'California'
    filtered_dutchbros_california = df_ipo[
        (df_ipo['Restaurant Chain'] == 'Dutch Bros') &
        (df_ipo['State'] == 'California')
    ]

    # Get unique ipo_base_links for the filtered chain
    links = filtered_dutchbros_california['ipo_base_link'].dropna().unique()

    print(f"--- Dutch Bros in California ({len(links)} links) ---")
    if len(links) > 0:
        for link in links:
            print(link)
    else:
        print("No links found for Dutch Bros in California.")
    print("\n")
else:
    print("Cannot extract links as df_ipo could not be loaded.")


Extracting ipo_base_links for Dutch Bros in California from df_ipo:

--- Dutch Bros in California (16 links) ---
https://www.yelp.com/biz/dutch-bros-coffee-fresno-7
https://www.yelp.com/biz/dutch-bros-coffee-fresno-2
https://www.yelp.com/biz/dutch-bros-coffee-roseville-2
https://www.yelp.com/biz/dutch-bros-coffee-paradise
https://www.yelp.com/biz/dutch-bros-coffee-gridley-2
https://www.yelp.com/biz/dutch-bros-coffee-san-jacinto-2
https://www.yelp.com/biz/dutch-bros-covina-2
https://www.yelp.com/biz/dutch-bros-coffee-reedley
https://www.yelp.com/biz/dutch-bros-coffee-brentwood-3
https://www.yelp.com/biz/dutch-bros-coffee-cathedral-city-2
https://www.yelp.com/biz/dutch-bros-coffee-williams
https://www.yelp.com/biz/dutch-bros-coffee-riverside-2
https://www.yelp.com/biz/dutch-bros-coffee-fresno-18
https://www.yelp.com/biz/dutch-bros-coffee-highland
https://www.yelp.com/biz/dutch-bros-coffee-sacramento-21
https://www.yelp.com/biz/dutch-bros-coffee-lincoln




**Reasoning**:
To determine the count of restaurant names for the common links, I will first filter `df_yelp` based on the `yelp_base_link` column and the `common_links` Series. Then, I will extract the `Restaurant` names from this filtered DataFrame. Finally, I will display the unique restaurant names and their counts.

In [ ]:
# Filter df_yelp to get only the rows corresponding to common links
common_yelp_restaurants = df_yelp[df_yelp['yelp_base_link'].isin(common_links)]

# Get the restaurant names and count their occurrences
restaurant_name_counts = common_yelp_restaurants['Restaurant'].value_counts().reset_index()
restaurant_name_counts.columns = ['Restaurant Name', 'Count']

print("Restaurant names and their counts from common Yelp links:")
display(restaurant_name_counts)

Restaurant names and their counts from common Yelp links:


,Restaurant Name,Count
0,"First Watch – League City, TX (1720 W FM 646 S...",1
1,"GEN Korean BBQ House – Houston, TX (18299 Egre...",1
2,"Portillo’s – Richmond, TX (11005 W Grand Pkwy S)",1
3,"CAVA – Anaheim, CA (8162 E Santa Ana Canyon Rd...",1


**Reasoning**:
To get the count for each restaurant name from the `pending_ipo_restaurants_dec2.csv` dataset, I will access the `df_ipo` DataFrame and use the `value_counts()` method on the 'Restaurant Chain' column. This will provide a count of how many times each restaurant chain appears in this dataset.

In [ ]:
# Get the counts of each restaurant chain in df_ipo
ipo_restaurant_counts = df_ipo['Restaurant Chain'].value_counts().reset_index()
ipo_restaurant_counts.columns = ['Restaurant Chain', 'Count']

print("Counts of each restaurant chain in pending_ipo_restaurants_dec2.csv:")
display(ipo_restaurant_counts)

Counts of each restaurant chain in pending_ipo_restaurants_dec2.csv:


,Restaurant Chain,Count
0,Dutch Bros,86
1,Sweet Green,66
2,Firstwatch,58
3,Portillo's,11
4,Reborn Coffee,11
5,Yoshiharu Ramen,11
6,Fatburger,10
7,CAVA,5
8,Krispy Kreme,5
9,Sweetgreen,4


# Task
The previous task focused on identifying common Yelp links between the two datasets. Now, I will proceed with standardizing the address information in both DataFrames to find common addresses and ultimately identify entries that match on both links and addresses.

Here's the plan:
*   **Standardize Addresses**: Clean and standardize the address columns in both DataFrames ('Location' in `df_yelp` and 'Address (Official)' in `df_ipo`) to facilitate accurate comparison. This will involve removing extra text, standardizing state abbreviations, and handling different address formats.
*   **Identify Common Addresses**: Find common addresses based on the standardized address columns.
*   **Identify Common Entries (Links and Addresses)**: Find entries that match on both standardized links and standardized addresses.
*   **Display and Summarize Matches**: Display the common addresses found and common entries based on both criteria. Provide a summary of the counts for each.

First, I will create a function to clean and standardize address strings. This function will convert addresses to lowercase, remove punctuation, remove common address components like 'suite' or 'road' (and their abbreviations), replace full state names with abbreviations (using the `State` column from `df_ipo`), and consolidate multiple spaces. Then, I will apply this function to create new 'cleaned_address' columns in both `df_yelp` and `df_ipo`.

Next, I will find the common cleaned addresses by taking the intersection of the unique 'cleaned_address' values from both DataFrames.

Finally, I will merge `df_yelp` and `df_ipo` using both the 'yelp_base_link' (or 'ipo_base_link') and 'cleaned_address' columns to identify rows that match on both criteria, providing a comprehensive view of the overlap between the datasets.
## Standardize Addresses

### Subtask:
Clean and standardize the address columns ('Location' in `df_yelp` and 'Address (Official)' in `df_ipo`) to prepare them for comparison.

### Reasoning:
To accurately compare addresses across two different datasets, it's crucial to standardize their format. This involves making them case-insensitive, removing irrelevant details like suite numbers or common street designators, and ensuring consistent state abbreviations. A custom function `clean_and_standardize_address` will be applied to both `df_yelp['Location']` and `df_ipo['Address (Official']`, leveraging `df_ipo['State']` for state abbreviation consistency.

### Code:
```python
import re
import pandas as pd

# State abbreviations mapping
state_abbreviations = {
    'alabama': 'al', 'alaska': 'ak', 'arizona': 'az', 'arkansas': 'ar', 'california': 'ca',
    'colorado': 'co', 'connecticut': 'ct', 'delaware': 'de', 'florida': 'fl', 'georgia': 'ga',
    'hawaii': 'hi', 'idaho': 'id', 'illinois': 'il', 'indiana': 'in', 'iowa': 'ia',
    'kansas': 'ks', 'kentucky': 'ky', 'louisiana': 'la', 'maine': 'me', 'maryland': 'md',
    'massachusetts': 'ma', 'michigan': 'mi', 'minnesota': 'mn', 'mississippi': 'ms',
    'missouri': 'mo', 'montana': 'mt', 'nebraska': 'ne', 'nevada': 'nv', 'new hampshire': 'nh',
    'new jersey': 'nj', 'new mexico': 'nm', 'new york': 'ny', 'north carolina': 'nc',
    'north dakota': 'nd', 'ohio': 'oh', 'oklahoma': 'ok', 'oregon': 'or', 'pennsylvania': 'pa',
    'rhode island': 'ri', 'south carolina': 'sc', 'south dakota': 'sd', 'tennessee': 'tn',
    'texas': 'tx', 'utah': 'ut', 'vermont': 'vt', 'virginia': 'va', 'washington': 'wa',
    'west virginia': 'wv', 'wisconsin': 'wi', 'wyoming': 'wy', 'district of columbia': 'dc'
}

def clean_and_standardize_address(address_str, full_state_name=None):
    """
    Cleans and standardizes an address string for comparison.
    Converts to lowercase, removes common address patterns (suites, street types),
    replaces full state names with abbreviations, removes punctuation, and normalizes whitespace.
    """
    if pd.isna(address_str):
        return None

    address_str = str(address_str).lower()

    # Remove common patterns like suite/apartment numbers (e.g., "suite 100", "ste b")
    address_str = re.sub(r'\b(suite|ste|apt|unit|floor|bldg)\s+[\w\d]+', '', address_str)
    # Remove common address designators (e.g., "road", "rd", "street", "st", "avenue", "ave")
    address_str = re.sub(r'\b(road|rd|street|st|avenue|ave|boulevard|blvd|drive|dr|lane|ln|court|ct|place|pl|terrace|ter|way|circle|cir|square|sq)\b', '', address_str)

    # Replace full state names with abbreviations if `full_state_name` is provided (for df_ipo)
    if full_state_name and isinstance(full_state_name, str):
        state_lower = full_state_name.lower()
        if state_lower in state_abbreviations:
            abbr = state_abbreviations[state_lower]
            # Replace the full state name with its abbreviation if found in the address string
            address_str = re.sub(r'\b' + re.escape(state_lower) + r'\b', abbr, address_str)

    # Remove punctuation, but keep numbers and letters
    address_str = re.sub(r'[^\w\s]', '', address_str)

    # Replace multiple spaces with a single space and strip leading/trailing spaces
    address_str = re.sub(r'\s+', ' ', address_str).strip()

    return address_str

# Apply to df_yelp
df_yelp['cleaned_address'] = df_yelp['Location'].apply(lambda x: clean_and_standardize_address(x))

# Apply to df_ipo, passing the State column for standardization
df_ipo['cleaned_address'] = df_ipo.apply(lambda row: clean_and_standardize_address(row['Address (Official)'], row['State']), axis=1)

print("df_yelp with new 'cleaned_address' column:")
print(df_yelp[['Location', 'cleaned_address']].head())
print("\n")
print("df_ipo with new 'cleaned_address' column:")
print(df_ipo[['Address (Official)', 'State', 'cleaned_address']].head())
```

## Standardize Addresses

### Subtask:
Clean and standardize the address columns ('Location' in `df_yelp` and 'Address (Official)' in `df_ipo`) to prepare them for comparison.


**Reasoning**:
To standardize the address columns for comparison, I will first define a dictionary for state abbreviations and then create a function to clean and normalize address strings, applying this function to the relevant address columns in both DataFrames while incorporating state abbreviation logic.



In [ ]:
import re

# 1. Define the state_abbreviations dictionary
state_abbreviations = {
    'Alabama': 'AL', 'Alaska': 'AK', 'Arizona': 'AZ', 'Arkansas': 'AR', 'California': 'CA',
    'Colorado': 'CO', 'Connecticut': 'CT', 'Delaware': 'DE', 'Florida': 'FL', 'Georgia': 'GA',
    'Hawaii': 'HI', 'Idaho': 'ID', 'Illinois': 'IL', 'Indiana': 'IN', 'Iowa': 'IA',
    'Kansas': 'KS', 'Kentucky': 'KY', 'Louisiana': 'LA', 'Maine': 'ME', 'Maryland': 'MD',
    'Massachusetts': 'MA', 'Michigan': 'MI', 'Minnesota': 'MN', 'Mississippi': 'MS',
    'Missouri': 'MO', 'Montana': 'MT', 'Nebraska': 'NE', 'Nevada': 'NV', 'New Hampshire': 'NH',
    'New Jersey': 'NJ', 'New Mexico': 'NM', 'New York': 'NY', 'North Carolina': 'NC',
    'North Dakota': 'ND', 'Ohio': 'OH', 'Oklahoma': 'OK', 'Oregon': 'OR', 'Pennsylvania': 'PA',
    'Rhode Island': 'RI', 'South Carolina': 'SC', 'South Dakota': 'SD', 'Tennessee': 'TN',
    'Texas': 'TX', 'Utah': 'UT', 'Vermont': 'VT', 'Virginia': 'VA', 'Washington': 'WA',
    'West Virginia': 'WV', 'Wisconsin': 'WI', 'Wyoming': 'WY',
    'District of Columbia': 'DC' # Added DC for completeness
}

# 2. Define the clean_and_standardize_address function
def clean_and_standardize_address(address, full_state_name=None):
    if pd.isna(address):
        return None
    address = str(address).lower()

    # Remove common patterns like suite/apartment numbers, street designators
    address = re.sub(r'suite\s\S+', '', address)
    address = re.sub(r'ste\s\S+', '', address)
    address = re.sub(r'apt\s\S+', '', address)
    address = re.sub(r'#\S+', '', address)
    address = re.sub(r'unit\s\S+', '', address)
    address = re.sub(r'bldg\s\S+', '', address)

    # Replace full street names with abbreviations or remove them if irrelevant
    address = re.sub(r'street', 'st', address)
    address = re.sub(r'avenue', 'ave', address)
    address = re.sub(r'road', 'rd', address)
    address = re.sub(r'boulevard', 'blvd', address)
    address = re.sub(r'drive', 'dr', address)
    address = re.sub(r'lane', 'ln', address)
    address = re.sub(r'parkway', 'pkwy', address)
    address = re.sub(r'court', 'ct', address)
    address = re.sub(r'terrace', 'ter', address)
    address = re.sub(r'place', 'pl', address)
    address = re.sub(r'north', 'n', address)
    address = re.sub(r'south', 's', address)
    address = re.sub(r'east', 'e', address)
    address = re.sub(r'west', 'w', address)

    # Replace full state names with abbreviations using the provided dictionary
    if full_state_name and full_state_name in state_abbreviations:
        address = address.replace(full_state_name.lower(), state_abbreviations[full_state_name].lower())
    else:
        # Also try to replace state names if they appear directly in the address string
        for state, abbr in state_abbreviations.items():
            address = address.replace(state.lower(), abbr.lower())

    # Remove punctuation
    address = re.sub(r'[.,-]', '', address)

    # Normalize whitespace (replace multiple spaces with a single space, strip leading/trailing spaces)
    address = re.sub(r'\s+', ' ', address).strip()

    return address

# 3. Apply the function to df_yelp['Location']
df_yelp['cleaned_address'] = df_yelp['Location'].apply(clean_and_standardize_address)

# 4. Apply the function to df_ipo['Address (Official)'], passing the corresponding 'State' value
df_ipo['cleaned_address'] = df_ipo.apply(lambda row: clean_and_standardize_address(row['Address (Official)'], row['State']), axis=1)

# 5. Print the head of df_yelp
print("\ndf_yelp with 'Location' and 'cleaned_address':")
print(df_yelp[['Location', 'cleaned_address']].head())

# 6. Print the head of df_ipo
print("\ndf_ipo with 'Address (Official)', 'State', and 'cleaned_address':")
print(df_ipo[['Address (Official)', 'State', 'cleaned_address']].head())


df_yelp with 'Location' and 'cleaned_address':
                     Location            cleaned_address
0           Webster, TX 77598           webster tx 77598
1       League City, TX 77573       league city tx 77573
2           Houston, TX 77058           houston tx 77058
3          Richmond, TX 77407          richmond tx 77407
4  Huntington Beach, CA 92646  huntington beach ca 92646

df_ipo with 'Address (Official)', 'State', and 'cleaned_address':
                                  Address (Official)       State  \
0  8162 E Santa Ana Canyon Road Suite 100 Anaheim...  California   
1    5100 Belt Line Road Suite 1056 Dallas, TX 75254       Texas   
2  5871 Long Prairie Road Suite 300 Flower Mound,...       Texas   
3  17603 La Cantera Parkway Building 100 San Anto...       Texas   
4     10850 Louetta Road Suite 100 Houston, TX 77070       Texas   

                                     cleaned_address  
0        8162 e santa ana canyon rd anaheim ca 92808  
1                  5100 

**Reasoning**:
Now that the address columns have been cleaned and standardized, I will identify the common addresses by finding the intersection of unique 'cleaned_address' values from both `df_yelp` and `df_ipo`. Then, I will print the count of these common addresses and the first few examples to verify the results.



In [ ]:
common_addresses = pd.Series(list(set(df_yelp['cleaned_address'].dropna()) & set(df_ipo['cleaned_address'].dropna())))

print(f"Number of common addresses found: {len(common_addresses)}")

if not common_addresses.empty:
    print("\nFirst 5 common addresses:")
    print(common_addresses.head())
else:
    print("\nNo common addresses found.")

Number of common addresses found: 0

No common addresses found.


**Reasoning**:
Since the individual common addresses yielded no results, I will now identify common entries by merging `df_yelp` and `df_ipo` based on both the 'yelp_base_link'/'ipo_base_link' and 'cleaned_address' columns. This will provide a comprehensive view of the overlap, even if it's empty, and aligns with the overall task of finding matches on both criteria.



In [ ]:
merged_df = pd.merge(
    df_yelp,
    df_ipo,
    left_on=['yelp_base_link', 'cleaned_address'],
    right_on=['ipo_base_link', 'cleaned_address'],
    how='inner'
)

print(f"Number of common entries (links and addresses) found: {len(merged_df)}")

if not merged_df.empty:
    print("\nFirst 5 common entries (links and addresses):")
    print(merged_df.head())
else:
    print("\nNo common entries found based on both links and addresses.")

Number of common entries (links and addresses) found: 0

No common entries found based on both links and addresses.


### Summary of Common Addresses and Combined Matches

Based on the standardization and comparison efforts:

*   **Common Addresses:** No common addresses were found between the `df_yelp` and `df_ipo` datasets after cleaning and standardizing the 'Location' and 'Address (Official)' columns. This indicates that even with rigorous cleaning, there were no exact matches in the textual representation of the addresses across the two datasets.

*   **Common Entries (Links and Addresses):** Similarly, when attempting to merge `df_yelp` and `df_ipo` based on both the standardized Yelp links (`yelp_base_link` / `ipo_base_link`) and the `cleaned_address` columns, **0 common entries** were identified. This means that no records shared both a matching Yelp base link and a matching cleaned address.

This outcome suggests that while there were 4 common Yelp links found in an earlier step, the associated address details, even after standardization, are not identical enough to produce matches when combined with the links. This could be due to subtle variations in address formatting, inclusion of different levels of detail (e.g., street number vs. full street address), or geographical differences in how addresses are recorded in each dataset for the same location.

## Final Task

### Subtask:
Provide a comprehensive summary of all findings, including the number of matching links and addresses, and discuss any insights from the comparison.


## Summary:

### Q&A
1.  **What was the outcome of comparing addresses and entries across the datasets?**
    Despite rigorous cleaning and standardization, no common addresses were found between the `df_yelp` and `df_ipo` datasets. Consequently, no common entries were identified when attempting to match on both Yelp links and cleaned addresses.

### Data Analysis Key Findings
*   A custom function, `clean_and_standardize_address`, was successfully applied to both `df_yelp['Location']` and `df_ipo['Address (Official)']` to create 'cleaned_address' columns. This process involved converting text to lowercase, removing common patterns like suite numbers and street designators, replacing full state names with abbreviations, removing punctuation, and normalizing whitespace.
*   After standardization, **0** common addresses were found between the two datasets, indicating no exact matches in the textual representation of addresses.
*   When attempting to identify entries matching on both standardized Yelp links (from a previous step) and the newly cleaned addresses, **0** common entries were found. This contrasts with a previous finding where **4** common Yelp links were identified.

### Insights or Next Steps
*   The absence of common addresses and combined link/address matches, despite the previous identification of common links, suggests significant discrepancies in address representation between the datasets. This could be due to variations in address detail (e.g., street number vs. full address), formatting nuances, or geographical recording differences that are not resolved by the current standardization.
*   Future steps could involve more advanced fuzzy matching techniques for addresses, breaking down addresses into components (street number, street name, city, zip) for comparison, or utilizing geocoding services to compare locations based on coordinates rather than text.


# Task
I cannot directly scrape review counts from Yelp due to the limitations of this environment's web scraping capabilities and external access restrictions.

Would you like to provide the review counts in another format, or is there a different task you'd like to perform with the existing data?

## Explain Limitations

### Subtask:
Inform the user that directly scraping review counts from Yelp is not possible due to limitations in web scraping capabilities and external access.


### Limitations on Yelp Review Scraping

It is important to note that directly scraping Yelp review counts from the platform is not feasible within this environment. This is due to inherent limitations in general web scraping capabilities and restrictions on external access, which prevent direct retrieval of such data. Therefore, we cannot proceed with extracting live review counts from Yelp.

## Final Task

### Subtask:
Ask the user if there's an alternative way to provide the review counts or if they have another task in mind that can be performed with the existing data.


## Summary:

### Q&A
Directly scraping Yelp review counts is not possible due to inherent limitations in web scraping capabilities and restrictions on external access within this environment.

### Data Analysis Key Findings
*   It is not feasible to directly scrape Yelp review counts from the platform.
*   This limitation is caused by general web scraping capabilities and external access restrictions.

### Insights or Next Steps
*   The next step is to await user input on providing review counts in an alternative format or suggesting a different task to perform with existing data.


# Task
Filter the `df_control_ca` DataFrame to include only outlets where the 'n_reviews' column is between 10 and 50 (exclusive of 10, inclusive of 50). Then, display the 'Outlet' name, 'url', and 'n_reviews' for all filtered entries, ensuring that outlet names are presented fully. Finally, summarize the findings, including the count of outlets found within this specified review range and list some example links.

## Filter Outlets by Review Count

### Subtask:
Filter the 'df_control_ca' DataFrame to include only outlets where the 'n_reviews' column is greater than 10 and less than or equal to 50.


**Reasoning**:
To filter the DataFrame according to the criteria, I will select rows where 'n_reviews' is greater than 10 and less than or equal to 50, storing the result in a new DataFrame.



In [1]:
filtered_outlets_10_to_50 = df_control_ca[
    (df_control_ca['n_reviews'] > 10) &
    (df_control_ca['n_reviews'] <= 50)
]

print("Outlets with review counts between 11 and 50 (inclusive):")
print(filtered_outlets_10_to_50[['Outlet', 'url', 'n_reviews']].head())
print(f"\nTotal outlets in this range: {len(filtered_outlets_10_to_50)}")

NameError: name 'df_control_ca' is not defined

**Reasoning**:
The previous cell failed because `df_control_ca` was not defined in the current execution environment. I will re-load `df_control_ca` from the provided CSV file, as it was done in a previous successful step, to ensure the DataFrame is available for filtering.



In [2]:
import pandas as pd

# Ensure df_control_ca is loaded
if 'df_control_ca' not in locals() or df_control_ca.empty:
    try:
        df_control_ca = pd.read_csv('/content/control_restaurants_CA_dec2025.csv')
        print("df_control_ca loaded successfully.")
    except FileNotFoundError:
        print("Error: '/content/control_restaurants_CA_dec2025.csv' not found. Cannot proceed.")
        df_control_ca = pd.DataFrame() # Create an empty DataFrame to avoid further errors

if not df_control_ca.empty:
    filtered_outlets_10_to_50 = df_control_ca[
        (df_control_ca['n_reviews'] > 10) &
        (df_control_ca['n_reviews'] <= 50)
    ]

    print("Outlets with review counts between 11 and 50 (inclusive):")
    print(filtered_outlets_10_to_50[['Outlet', 'url', 'n_reviews']].head())
    print(f"\nTotal outlets in this range: {len(filtered_outlets_10_to_50)}")
else:
    print("Cannot filter as df_control_ca could not be loaded.")

Error: '/content/control_restaurants_CA_dec2025.csv' not found. Cannot proceed.
Cannot filter as df_control_ca could not be loaded.


**Reasoning**:
The previous attempt to load `df_control_ca` failed because the `.csv` file was not found. Based on the selected files, the correct file is likely `/content/control_restaurants_CA_dec2025_matched_green.xlsx`. I will update the code to load this Excel file instead, ensuring the DataFrame is correctly initialized before filtering.



In [3]:
import pandas as pd

# Ensure df_control_ca is loaded
if 'df_control_ca' not in locals() or df_control_ca.empty:
    try:
        # Attempt to load the Excel file
        df_control_ca = pd.read_excel('/content/control_restaurants_CA_dec2025_matched_green.xlsx')
        print("df_control_ca loaded successfully from Excel file.")
    except FileNotFoundError:
        print("Error: '/content/control_restaurants_CA_dec2025_matched_green.xlsx' not found. Cannot proceed.")
        df_control_ca = pd.DataFrame() # Create an empty DataFrame to avoid further errors

if not df_control_ca.empty:
    filtered_outlets_10_to_50 = df_control_ca[
        (df_control_ca['n_reviews'] > 10) &
        (df_control_ca['n_reviews'] <= 50)
    ]

    print("Outlets with review counts between 11 and 50 (inclusive):")
    print(filtered_outlets_10_to_50[['Outlet', 'url', 'n_reviews']].head())
    print(f"\nTotal outlets in this range: {len(filtered_outlets_10_to_50)}")
else:
    print("Cannot filter as df_control_ca could not be loaded.")

df_control_ca loaded successfully from Excel file.
Outlets with review counts between 11 and 50 (inclusive):
                      Outlet  \
0              Aromas Coffee   
4   Crenshaw Farmerss Market   
11                Donut Star   
16                Donut Star   
18                Donut Star   

                                                  url  n_reviews  
0   https://www.yelp.com/biz/aromas-coffee-los-ang...         24  
4   https://www.yelp.com/biz/crenshaw-farmers-mark...         27  
11      https://www.yelp.com/biz/donut-star-mira-loma         13  
16    https://www.yelp.com/biz/donut-star-westminster         31  
18        https://www.yelp.com/biz/donut-star-anaheim         41  

Total outlets in this range: 289


## Display Filtered Links

### Subtask:
Display the 'Outlet' name, 'url', and 'n_reviews' for all filtered entries, ensuring that outlet names are presented fully.


**Reasoning**:
To display all filtered entries as requested, I will print a descriptive header and then use the `to_string(index=False)` method on the `filtered_outlets_10_to_50` DataFrame, selecting the 'Outlet', 'url', and 'n_reviews' columns to ensure full display of outlet names and all rows without an index.



In [5]:
print("All outlets with review counts between 11 and 50 (sorted by number of reviews ascending):\n")

if not filtered_outlets_10_to_50.empty:
    # Sort the DataFrame by 'n_reviews' in ascending order
    sorted_filtered_outlets = filtered_outlets_10_to_50.sort_values(by='n_reviews', ascending=True)
    print(sorted_filtered_outlets[['Outlet', 'url', 'n_reviews']].to_string(index=False))
else:
    print("No outlets found with review counts between 11 and 50.")

All outlets with review counts between 11 and 50 (sorted by number of reviews ascending):

                        Outlet                                                                              url  n_reviews
                        Subway                                            https://www.yelp.com/biz/subway-linda         11
                        Subway                                       https://www.yelp.com/biz/subway-manteca-13         11
                        Subway                                      https://www.yelp.com/biz/subway-stockton-45         11
                The Happy Inka                              https://www.yelp.com/biz/the-happy-inka-los-angeles         11
                    Tower Cafe                                       https://www.yelp.com/biz/tower-cafe-encino         11
        Krispy Krunchy Chicken                       https://www.yelp.com/biz/krispy-krunchy-chicken-berkeley-2         11
                        Subway                  

In [4]:
print("All outlets with review counts between 51 and 55 (sorted by number of reviews ascending):\n")

if not filtered_outlets_51_to_55.empty:
    # Sort the DataFrame by 'n_reviews' in ascending order
    sorted_filtered_outlets_51_55 = filtered_outlets_51_to_55.sort_values(by='n_reviews', ascending=True)
    print(sorted_filtered_outlets_51_55[['Outlet', 'url', 'n_reviews']].to_string(index=False))
else:
    print("No outlets found with review counts between 51 and 55.")

All outlets with review counts between 51 and 55 (sorted by number of reviews ascending):

                Outlet                                                                               url  n_reviews
                Subway                                      https://www.yelp.com/biz/subway-escondido-24         51
                Subway                                           https://www.yelp.com/biz/subway-norco-8         51
                Subway                                      https://www.yelp.com/biz/subway-roseville-41         51
                Subway                                      https://www.yelp.com/biz/subway-sunnyvale-21         52
        Tous Les Jours                                  https://www.yelp.com/biz/tous-les-jours-irvine-5         52
                Subway                                        https://www.yelp.com/biz/subway-ontario-46         52
                Subway                               https://www.yelp.com/biz/subway-huntington-b

## Filter Outlets by Review Count (56-60)

### Subtask:
Filter the 'df_control_ca' DataFrame to include only outlets where the 'n_reviews' column is greater than 55 and less than or equal to 60.

**Reasoning**:
To filter the DataFrame for outlets with review counts between 56 and 60, I will select rows where 'n_reviews' is greater than 55 and less than or equal to 60. I'll then store these results in a new DataFrame and display its head, along with the total count of outlets found in this range.

In [5]:
import pandas as pd

# Ensure df_control_ca is loaded
if 'df_control_ca' not in locals() or df_control_ca.empty:
    try:
        df_control_ca = pd.read_excel('/content/control_restaurants_CA_dec2025_highlighted_combined_v5.xlsx')
        print("df_control_ca loaded successfully from Excel file.")
    except FileNotFoundError:
        print("Error: '/content/control_restaurants_CA_dec2025_highlighted_combined_v5.xlsx' not found. Cannot proceed.")
        df_control_ca = pd.DataFrame() # Create an empty DataFrame to avoid further errors

if not df_control_ca.empty:
    filtered_outlets_56_to_60 = df_control_ca[
        (df_control_ca['n_reviews'] > 55) &
        (df_control_ca['n_reviews'] <= 60)
    ]

    print("Outlets with review counts between 56 and 60 (inclusive):\n")
    print(filtered_outlets_56_to_60[['Outlet', 'url', 'n_reviews']].head())
    print(f"\nTotal outlets in this range: {len(filtered_outlets_56_to_60)}")
else:
    print("Cannot filter as df_control_ca could not be loaded.")

Outlets with review counts between 56 and 60 (inclusive):

     Outlet                                            url  n_reviews
120  Subway      https://www.yelp.com/biz/subway-upland-21         60
133  Subway     https://www.yelp.com/biz/subway-la-jolla-7         58
137  Subway  https://www.yelp.com/biz/subway-san-diego-197         58
142  Subway  https://www.yelp.com/biz/subway-san-diego-218         58
143  Subway  https://www.yelp.com/biz/subway-san-diego-209         58

Total outlets in this range: 20


## Display Filtered Links (56-60)

### Subtask:
Display the 'Outlet' name, 'url', and 'n_reviews' for all filtered entries (n_reviews between 56 and 60), ensuring that outlet names are presented fully and sorted by the number of reviews.

**Reasoning**:
To display all filtered entries as requested, I will first sort the `filtered_outlets_56_to_60` DataFrame by the 'n_reviews' column in ascending order. Then, I will print a descriptive header and use the `to_string(index=False)` method to ensure full display of outlet names and all rows without an index, selecting the 'Outlet', 'url', and 'n_reviews' columns.

In [6]:
print("All outlets with review counts between 56 and 60 (sorted by number of reviews ascending):\n")

if not filtered_outlets_56_to_60.empty:
    # Sort the DataFrame by 'n_reviews' in ascending order
    sorted_filtered_outlets_56_60 = filtered_outlets_56_to_60.sort_values(by='n_reviews', ascending=True)
    print(sorted_filtered_outlets_56_60[['Outlet', 'url', 'n_reviews']].to_string(index=False))
else:
    print("No outlets found with review counts between 56 and 60.")

All outlets with review counts between 56 and 60 (sorted by number of reviews ascending):

                      Outlet                                                                 url  n_reviews
                      Subway                           https://www.yelp.com/biz/subway-tustin-23         56
                      Subway                     https://www.yelp.com/biz/subway-rancho-mirage-4         56
                      Subway                   https://www.yelp.com/biz/subway-citrus-heights-20         56
                      Subway                      https://www.yelp.com/biz/subway-santa-clara-18         56
                      Subway                      https://www.yelp.com/biz/subway-sacramento-169         56
                      Subway                        https://www.yelp.com/biz/subway-daly-city-17         57
              Yum Yum Donuts                   https://www.yelp.com/biz/yum-yum-donuts-la-mirada         57
                      Subway                 

## Filter Outlets by Review Count (61-70)

### Subtask:
Filter the 'df_control_ca' DataFrame to include only outlets where the 'n_reviews' column is greater than 60 and less than or equal to 70.

**Reasoning**:
To filter the DataFrame for outlets with review counts between 61 and 70, I will select rows where 'n_reviews' is greater than 60 and less than or equal to 70. I'll then store these results in a new DataFrame and display its head, along with the total count of outlets found in this range.

In [1]:
import pandas as pd

# Ensure df_control_ca is loaded
if 'df_control_ca' not in locals() or df_control_ca.empty:
    try:
        df_control_ca = pd.read_excel('/content/control_restaurants_CA_dec2025_highlighted_combined_v5.xlsx')
        print("df_control_ca loaded successfully from Excel file.")
    except FileNotFoundError:
        print("Error: '/content/control_restaurants_CA_dec2025_highlighted_combined_v5.xlsx' not found. Cannot proceed.")
        df_control_ca = pd.DataFrame() # Create an empty DataFrame to avoid further errors

if not df_control_ca.empty:
    filtered_outlets_61_to_70 = df_control_ca[
        (df_control_ca['n_reviews'] > 60) &
        (df_control_ca['n_reviews'] <= 70)
    ]

    print("Outlets with review counts between 61 and 70 (inclusive):\n")
    print(filtered_outlets_61_to_70[['Outlet', 'url', 'n_reviews']].head())
    print(f"\nTotal outlets in this range: {len(filtered_outlets_61_to_70)}")
else:
    print("Cannot filter as df_control_ca could not be loaded.")

df_control_ca loaded successfully from Excel file.
Outlets with review counts between 61 and 70 (inclusive):

              Outlet                                                url  \
13        Donut Star  https://www.yelp.com/biz/donut-star-lake-elsinore   
23        Donut Star       https://www.yelp.com/biz/donut-star-woodland   
83   New York Donuts  https://www.yelp.com/biz/new-york-donuts-stock...   
114           Subway    https://www.yelp.com/biz/subway-jurupa-valley-2   
117           Subway         https://www.yelp.com/biz/subway-ontario-51   

     n_reviews  
13          65  
23          62  
83          69  
114         64  
117         62  

Total outlets in this range: 33


## Display Filtered Links (61-70)

### Subtask:
Display the 'Outlet' name, 'url', and 'n_reviews' for all filtered entries (n_reviews between 61 and 70), ensuring that outlet names are presented fully and sorted by the number of reviews.

**Reasoning**:
To display all filtered entries as requested, I will first sort the `filtered_outlets_61_to_70` DataFrame by the 'n_reviews' column in ascending order. Then, I will print a descriptive header and use the `to_string(index=False)` method to ensure full display of outlet names and all rows without an index, selecting the 'Outlet', 'url', and 'n_reviews' columns.

In [2]:
print("All outlets with review counts between 61 and 70 (sorted by number of reviews ascending):\n")

if not filtered_outlets_61_to_70.empty:
    # Sort the DataFrame by 'n_reviews' in ascending order
    sorted_filtered_outlets_61_70 = filtered_outlets_61_to_70.sort_values(by='n_reviews', ascending=True)
    print(sorted_filtered_outlets_61_70[['Outlet', 'url', 'n_reviews']].to_string(index=False))
else:
    print("No outlets found with review counts between 61 and 70.")

All outlets with review counts between 61 and 70 (sorted by number of reviews ascending):

                      Outlet                                                               url  n_reviews
The Coffee Bean and Tea Leaf https://www.yelp.com/biz/the-coffee-bean-and-tea-leaf-bakersfield         61
                      Subway                        https://www.yelp.com/biz/subway-antelope-2         61
                  Donut Star                      https://www.yelp.com/biz/donut-star-woodland         62
                      Subway                        https://www.yelp.com/biz/subway-ontario-51         62
                      Subway                       https://www.yelp.com/biz/subway-hesperia-13         62
                      Subway                    https://www.yelp.com/biz/subway-sacramento-177         62
                      Subway                       https://www.yelp.com/biz/subway-beaumont-33         62
                     Trimana                 https://www.yelp

## Summary of Outlets with Review Counts between 61 and 70

### Final Task:
Summarize the findings, including the count of outlets found within the specified review range (61-70) and list some example links.

## Summary of Outlets with Review Counts between 56 and 60

### Final Task:
Summarize the findings, including the count of outlets found within the specified review range (56-60) and list some example links.

In [3]:
import pandas as pd

# Ensure df_control_ca is loaded
if 'df_control_ca' not in locals() or df_control_ca.empty:
    try:
        df_control_ca = pd.read_excel('/content/control_restaurants_CA_dec2025_highlighted_combined_v5.xlsx')
        print("df_control_ca loaded successfully from Excel file.")
    except FileNotFoundError:
        print("Error: '/content/control_restaurants_CA_dec2025_highlighted_combined_v5.xlsx' not found. Cannot proceed.")
        df_control_ca = pd.DataFrame() # Create an empty DataFrame to avoid further errors

if not df_control_ca.empty:
    filtered_outlets_51_to_55 = df_control_ca[
        (df_control_ca['n_reviews'] > 50) &
        (df_control_ca['n_reviews'] <= 55)
    ]

    print("Outlets with review counts between 51 and 55 (inclusive):\n")
    print(filtered_outlets_51_to_55[['Outlet', 'url', 'n_reviews']].head())
    print(f"\nTotal outlets in this range: {len(filtered_outlets_51_to_55)}")
else:
    print("Cannot filter as df_control_ca could not be loaded.")

df_control_ca loaded successfully from Excel file.
Outlets with review counts between 51 and 55 (inclusive):

                     Outlet  \
15               Donut Star   
71   Krispy Krunchy Chicken   
105                  Subway   
106                  Subway   
116                  Subway   

                                                   url  n_reviews  
15          https://www.yelp.com/biz/donut-star-irvine         55  
71   https://www.yelp.com/biz/krispy-krunchy-chicke...         54  
105        https://www.yelp.com/biz/subway-valencia-19         53  
106            https://www.yelp.com/biz/subway-azusa-5         53  
116         https://www.yelp.com/biz/subway-ontario-46         52  

Total outlets in this range: 22


## Filter Outlets by Review Count (51-55)

### Subtask:
Filter the 'df_control_ca' DataFrame to include only outlets where the 'n_reviews' column is greater than 50 and less than or equal to 55.

**Reasoning**:
To filter the DataFrame for outlets with review counts between 51 and 55, I will select rows where 'n_reviews' is greater than 50 and less than or equal to 55. I'll then store these results in a new DataFrame and display its head, along with the total count of outlets found in this range.

In [1]:
import pandas as pd

# Ensure df_control_ca is loaded
if 'df_control_ca' not in locals() or df_control_ca.empty:
    try:
        df_control_ca = pd.read_excel('/content/control_restaurants_CA_dec2025_matched_green.xlsx')
        print("df_control_ca loaded successfully from Excel file.")
    except FileNotFoundError:
        print("Error: '/content/control_restaurants_CA_dec2025_matched_green.xlsx' not found. Cannot proceed.")
        df_control_ca = pd.DataFrame() # Create an empty DataFrame to avoid further errors

if not df_control_ca.empty:
    filtered_outlets_51_to_55 = df_control_ca[
        (df_control_ca['n_reviews'] > 50) &
        (df_control_ca['n_reviews'] <= 55)
    ]

    print("Outlets with review counts between 51 and 55 (inclusive):\n")
    print(filtered_outlets_51_to_55[['Outlet', 'url', 'n_reviews']].head())
    print(f"\nTotal outlets in this range: {len(filtered_outlets_51_to_55)}")
else:
    print("Cannot filter as df_control_ca could not be loaded.")

Error: '/content/control_restaurants_CA_dec2025_matched_green.xlsx' not found. Cannot proceed.
Cannot filter as df_control_ca could not be loaded.


## Display Filtered Links (51-55)

### Subtask:
Display the 'Outlet' name, 'url', and 'n_reviews' for all filtered entries (n_reviews between 51 and 55), ensuring that outlet names are presented fully and sorted by the number of reviews.

**Reasoning**:
To display all filtered entries as requested, I will first sort the `filtered_outlets_51_to_55` DataFrame by the 'n_reviews' column in ascending order. Then, I will print a descriptive header and use the `to_string(index=False)` method to ensure full display of outlet names and all rows without an index, selecting the 'Outlet', 'url', and 'n_reviews' columns.

In [2]:
print("All outlets with review counts between 51 and 55 (sorted by number of reviews ascending):\n")

if not filtered_outlets_51_to_55.empty:
    # Sort the DataFrame by 'n_reviews' in ascending order
    sorted_filtered_outlets_51_55 = filtered_outlets_51_to_55.sort_values(by='n_reviews', ascending=True)
    print(sorted_filtered_outlets_51_55[['Outlet', 'url', 'n_reviews']].to_string(index=False))
else:
    print("No outlets found with review counts between 51 and 55.")

All outlets with review counts between 51 and 55 (sorted by number of reviews ascending):



NameError: name 'filtered_outlets_51_to_55' is not defined

## Summary of Outlets with Review Counts between 51 and 55

### Final Task:
Summarize the findings, including the count of outlets found within the specified review range (51-55) and list some example links.

## Summary of Outlets with Review Counts between 11 and 50

A total of **289** outlets in California were found to have Yelp review counts between 11 and 50 (inclusive). This range represents businesses that have established a presence on Yelp with a moderate number of reviews, indicating more customer engagement than those with very few reviews, but not yet reaching the very high review counts of established, high-volume businesses.

### Example Links:

*   Aromas Coffee: https://www.yelp.com/biz/aromas-coffee-los-angeles
*   Crenshaw Farmerss Market: https://www.yelp.com/biz/crenshaw-farmers-market-los-angeles
*   Donut Star: https://www.yelp.com/biz/donut-star-mira-loma
*   Krispy Krunchy Chicken: https://www.yelp.com/biz/krispy-krunchy-chicken-san-bernardino
*   Subway: https://www.yelp.com/biz/subway-long-beach-45

## Final Task

### Subtask:
Summarize the findings, including the count of outlets found within the specified review range and list some example links.


## Summary:

### Q&A
*   **How many outlets were found within the specified review range (11 to 50 reviews)?**
    A total of 289 outlets were found to have Yelp review counts between 11 and 50 (inclusive).
*   **Can you list some example links for these outlets?**
    Some example links include:
    *   Aromas Coffee: `https://www.yelp.com/biz/aromas-coffee-los-angeles`
    *   Crenshaw Farmerss Market: `https://www.yelp.com/biz/crenshaw-farmers-market-los-angeles`
    *   Donut Star: `https://www.yelp.com/biz/donut-star-mira-loma`
    *   Krispy Krunchy Chicken: `https://www.yelp.com/biz/krispy-krunchy-chicken-san-bernardino`
    *   Subway: `https://www.yelp.com/biz/subway-long-beach-45`

### Data Analysis Key Findings
*   The `df_control_ca` DataFrame was successfully loaded from `control_restaurants_CA_dec2025_matched_green.xlsx` after initial attempts with incorrect file paths and types failed.
*   After filtering, 289 outlets were identified as having review counts between 11 and 50 (exclusive of 10, inclusive of 50).
*   For these 289 outlets, the 'Outlet' name, 'url', and 'n_reviews' were displayed, with outlet names fully presented.

### Insights or Next Steps
*   The identified outlets represent businesses with a moderate level of customer engagement on Yelp, making them potential targets for marketing or operational analysis.
*   Further analysis could involve categorizing these 289 outlets by cuisine type, location, or average rating to identify patterns within this specific engagement tier.
